ENH: special.rel_entr: Avoid overflow before computing log

nickodell opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Fixes #20710

#### What does this implement/fix?
This PR improves the accuracy of `rel_entr()` in two ways:

1. If x / y overflows, underflows, or results in a subnormal number, the result is computed with the formula `x * (log(x) - log(y))`.
2. If x / y is close to 1, then a formula based on log1p is used instead. This improves accuracy for small rel_entr() values.

#### Performance regression

This is slower than the original `rel_entr()`.

Example benchmark script:

```
from scipy.special import rel_entr
import numpy as np
import timeit

x = np.random.rand(1_000_000)
y = np.random.rand(1_000_000)

def func():
    rel_entr(x, y)

print(timeit.timeit(func, number=100))
```

Prints 1.958 on main vs 3.587 on my branch.

#### Additional information

The log1p formula was contributed by @fancidev. 

I didn't understand why it worked, so here's a derivation of the formula:

```
rel_entry(x, y) = x * log(x / y)
rel_entry(x, y) = x * log((x - y + y) / y)
rel_entry(x, y) = x * log(1 + (x - y) / y)
rel_entry(x, y) = x * log1p((x - y) / y)
```

cc @mdhaber

nickodell committed:
TST: Add tests for rel_entr overflow

Also add accuracy tests for it

nickodell committed:
ENH: special.rel_entr: avoid overflow

For very small y, and large x, the expression x / y can overflow. Avoid
this by doing the logarithm first if the division overflows. Also use
alternate formula for rel_entr when x and y are very close.

Co-Authored-By: fancidev <fancidev@gmail.com>

mdhaber commented:
Thanks, it would be great to avoid the over/underflow. I'll leave it to the `special` maintainers to review.

nickodell commented:
Windows CI failure seems to be related to gh-20806.

```
 ______________ TestDifferentialEvolutionSolver.test_strategy_fn _______________
[gw0] win32 -- Python 3.12.3 C:\hostedtoolcache\windows\Python\3.12.3\x64\python.exe
Test passed but took too long to run: Duration 1.8958849000000555s > 1.0s
```

Looking into the [CircleCI failure](https://app.circleci.com/pipelines/github/scipy/scipy/27523/workflows/38d267a8-03ac-4ab6-a4b6-091b0a4b16a1/jobs/92817) now.

nickodell commented:
Seems that this change to `rel_entr()` means that the doc test no longer passes. The new value differs by one ULP from the old value. Checking with mpmath, the new value seems to be closer to the true value.

Here's a possible fix. It is a bit ugly, though - I think it makes the docs a little less clear here. What do you think?

```
diff --git a/scipy/stats/_entropy.py b/scipy/stats/_entropy.py
index 1e6c8d9e1..79954e24c 100644
--- a/scipy/stats/_entropy.py
+++ b/scipy/stats/_entropy.py
@@ -123,8 +123,8 @@ def entropy(pk: np.typing.ArrayLike,
 
     >>> D = entropy(pk, qk, base=base)
     >>> D
-    0.7369655941662062
-    >>> D == np.sum(pk * np.log(pk/qk)) / np.log(base)
+    0.7369655941662061
+    >>> np.isclose(D, np.sum(pk * np.log(pk/qk)) / np.log(base))
     True
 
     The cross entropy can be calculated as the sum of the entropy and
```

fancidev commented:
> Seems that this change to `rel_entr()` means that the doc test for `scipy.stats.entropy()` no longer passes. The new value differs by one ULP from the old value. Checking with mpmath, the new value seems to be closer to the true value.

Nice catch. What is the correctly rounded result? (I find that by `float(x)` where `x` is the result produced by `mpmath`.)

fancidev commented:
> I did the same thing to round it to a float. Here's the unrounded result:
> 

Cool, this shows the new result is more accurate than the old one.

I don't know what the maintainers think, but I would remove those lines from the doc test:

```python
>>> D == np.sum(pk * np.log(pk/qk)) / np.log(base)
    True
```

For a few reasons. (1) Exact floating point equality should rarely be used for testing. (2) There is already unit tests for this function (I hope). (3) It doesn't appear to be best practice to test against a code that reproduces the expression.

mdhaber commented:
Just to clarify, the primary purpose of doctests here is not testing but to demonstrate the API to the user. We do run the checks, but mostly to make sure the syntax is right and that the example is showing what we think it is. There may be other ways of doing this without an equality sign (e.g. print both results), but sometimes that is the most obvious way to make the point.

fancidev commented:
The code looks good to me. (I also verified that the value of `tiny` is correctly copy-pasted.)

In addition, if `x` and `y` are both INFINITY, the function returns `nan` both before and after the PR. (Whether this is desirable/expected may be subject to discussion, but the PR retains the behaviour.)

nickodell commented:
@mdhaber I see you re-triggered the test - do you need me to merge main into this branch so that the fail-slow test passes?

steppi commented on scipy/special/_convex_analysis.pxd:
```diff
@@ -20,11 +21,26 @@ cdef inline double kl_div(double x, double y) noexcept nogil:
 cdef inline double rel_entr(double x, double y) noexcept nogil:
+    cdef double ratio
+    # Smallest non-subnormal number, aka np.finfo(np.float64).tiny
+    cdef double tiny = 2.2250738585072014e-308
```
You can get this value through

```cython
from libc.float cimport DBL_MIN
```

so no need to define it within the function.

nickodell replied:
Thanks, thought there must be a better way to do this.

Fixed in 78b66f8.

steppi commented on scipy/special/_convex_analysis.pxd:
```diff
@@ -20,11 +21,26 @@ cdef inline double kl_div(double x, double y) noexcept nogil:
+            return x * (log(x) - log(y))
     elif x == 0 and y >= 0:
         return 0
     else:
```
GitHub isn't letting me turn this into a suggestion, but I think it would be good to avoid most of the nested indentation. You could structure this as

```cython
    if isnan(x) or isnan(y):
        return NAN
    if x <= 0 or y <= 0:
        if x == 0 and y >= 0:
            return 0
        return INFINITY
    ratio = x / y
    if 0.5 < ratio < 2:
        # When x and y are close, this is more accurate
        return x * log1p((x - y) / y)
    if DBL_MIN < ratio < INFINITY:
        # There are no underflow/overflow issues
        return x * log(ratio)
    # x and y are so far apart that taking x / y
    # results in either an underflow, overflow,
    # or subnormal number. Do the logarithm first
    return x * (log(x) - log(y))
```

I replaced `tiny` with `DBL_MIN` following my other suggestion.

nickodell replied:
Fixed in 3727337.

steppi reviewed (changes requested):
Looks good @nickodell and I think the improved accuracy outweighs the increase in runtime. Just a few minor suggestions.

mdhaber commented:
I would need to merge main because it turns out that test got much slower recently. But you can also just ignore the failure. I was just trying to make CI green for you.

nickodell committed:
TST: Reduce absolute tolerance for rel_entr test

This change is required, because otherwise this test does not fail when
subnormals are not correctly handled.

nickodell committed:
ENH: Find smallest normal from C header

nickodell committed:
ENH: remove nested if from rel_entr

nickodell committed:
DOC: Remove docs which require exact float match

Because rel_entr can be calculated in one of two ways, the relative
entropy formula is no longer accurate for x and y which are similar to
one another. However, the numpy calculation is still a useful inuition
for what the calculation is doing.

steppi commented on scipy/stats/_entropy.py:
```diff
@@ -123,9 +123,9 @@ def entropy(pk: np.typing.ArrayLike,
-    0.7369655941662062
-    >>> D == np.sum(pk * np.log(pk/qk)) / np.log(base)
-    True
+    0.736965594166206
+    >>> np.sum(pk * np.log(pk/qk)) / np.log(base)
```
How about using the actual value that you're getting for `D to avoid the "slight lie" and then using `np.allclose` with a very small `rtol` (around twice machine epsilon?).

nickodell replied:
See 3da2b2d.

nickodell committed:
Revert "DOC: Remove docs which require exact float match"

This reverts commit d4e910ca85e9706d6810786f527e59adc4d73ff9.

nickodell committed:
DOC: Don't require exact equivalence in entropy docs

[docs only]

nickodell committed:
DOC: Tighten relative tolerance allowed for entropy example

[docs only]

steppi reviewed (approved):
Looks good. Thanks @nickodell!

WarrenWeckesser commented on scipy/special/_convex_analysis.pxd:
```diff
@@ -20,15 +22,26 @@ cdef inline double kl_div(double x, double y) noexcept nogil:
-    elif x == 0 and y >= 0:
-        return 0
-    else:
+    if x <= 0 or y <= 0:
+        if x == 0 and y >= 0:
+            return 0
         return INFINITY
```
As long as you are adding code to get the edge cases correct, you might as well also handle x = ±∞ and y = ±∞.  In all four possibilities, the function should return nan.

steppi replied:
My understanding is that handling out of domain cases by returning $\infty$ (or $-\infty$ for concave functions) was a deliberate choice in `_convex_analysis.pxd`, and this behavior is consistent across the functions defined in this file. See for instance the discussion in https://github.com/scipy/scipy/issues/3421. If we should change this behavior, I think it needs it's own discussion first, and if we proceed, the change should be done simultaneously for all functions in `_convex_analysis.pxd`. It shouldn't impact this PR.

One thing that may still be inconsistent is that when `x` and `y` are both $+\infty$, the output is `nan`, as @fancidev pointed out above. I think this can be addressed separately if needed.

WarrenWeckesser replied:
@steppi, I see, thanks.  I hadn't followed that discussion closely, so I was applying the usual rules that `inf/inf` and `inf - inf` return `nan`.

steppi commented:
@WarrenWeckesser, I've merged this but I'll create an issue today about edge case behavior for these convex analysis functions.
